rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp

```diff
@@ -53,11 +55,26 @@ enum class publishing_mode_t
   AUTO           // Use publishing mode set in XML file or Fast DDS default
 };
 
+typedef struct UseCountTopic
+{
+  eprosima::fastdds::dds::Topic * topic{nullptr};
+  size_t use_count{0};
+} UseCountTopic;
+
 typedef struct CustomParticipantInfo
 {
   eprosima::fastdds::dds::DomainParticipant * participant_{nullptr};
   ParticipantListener * listener_{nullptr};
 
+  std::mutex topic_name_to_topic_mutex_;
+  // As of 2023-02-07, Fast-DDS only allows one create_topic() with the same
+  // topic name per DomainParticipant.  Thus, we need to check if the topic
+  // was already created.  If it did, then we just increase the use_count
+  // that we are tracking, and return the existing topic.  If it
+  // didn't, then we create a new one and start tracking it.  Once all
+  // users of the topic are removed, we will delete the topic.
+  std::map<std::string, std::unique_ptr<UseCountTopic>> topic_name_to_topic_;
```

Comment:
We should have a destructor that iterates through the `topic_name_to_topic_` map and calls `participant_->delete_topic()` on each, so we don't leak

Reply:
That's right, `destroy_participant` does this.  That is `destroy_participant` calls `remove_topic_and_type` on every type that is currently registered, and in turn `remove_topic_and_type` calls `CustomParticipantInfo::delete_topic` for every topic.  So we should be good here.

(I double-checked this by running under valgrind, and I didn't see any new leaks with this in place)